AGILE/Commands.cs

```diff
@@ -2103,18 +2085,94 @@ public byte ExecuteLogic(int logicNum)
 
             // Restore the previous Logic number before we leave.
             state.CurrentLogNum = previousLogNum;
 
-            // If new.room was not one of the Actions executed, then newRoom will still have the current
-            // room value; otherwise it will have the number of the new room.
-            return newRoom;
+            // If ExecuteAction return 0, then it means that a newroom, restore or restart is 
+            // happening. In those cases, we need to immediately rescan logics from the top of Logic.0
+            return (actionNum == 0);
+        }
+
+        /// <summary>
+        /// Performs all the necessary updates to vars, flags, animated objects, controllers, 
+        /// and other state to prepare for entry in to the next room.
+        /// </summary>
+        /// <param name="roomNum"></param>
+        private void NewRoom(byte roomNum)
+        {
+            // Simulate a slow room change if there is a text window open.
+            if (textGraphics.IsWindowOpen()) Thread.Sleep(1000);
+
+            // Turn off sound.
+            soundPlayer.Reset();
+
+            // Clear the script event buffer ready for next room.
+            state.ScriptBuffer.InitScript();
+            state.ScriptBuffer.ScriptOn();
+
+            // Resets the Logics, Views, Pictures and Sounds back to new room state.
+            state.ResetResources();
+
+            // Carry over ego's view number.
+            // TODO: For some reason in MH2, the ego View can be null at this point. Needs investigation to determine why.
+            if (state.Ego.View != null)
+            {
+                state.Vars[Defines.CURRENT_EGO] = (byte)state.Ego.View.Index;
+            }
+
+            // Reset state for all animated objects.
+            foreach (AnimatedObject aniObj in state.AnimatedObjects) aniObj.Reset();
+
+            // Current room logic is loaded automatically on room change and not directly by load.logic
+            Logic logic = state.Logics[roomNum];
+            logic.IsLoaded = true;
+            state.ScriptBuffer.AddScript(ScriptBuffer.ScriptBufferEventType.LoadLogic, logic.Index);
+
+            // If ego collided with a border, set his position in the new room to
+            // the appropriate edge of the screen.
+            switch (state.Vars[Defines.EGOEDGE])
+            {
+                case Defines.TOP:
+                    state.Ego.Y = Defines.MAXY;
+                    break;
+
+                case Defines.RIGHT:
+                    state.Ego.X = Defines.MINX;
+                    break;
+
+                case Defines.BOTTOM:
+                    state.Ego.Y = Defines.HORIZON + 1;
+                    break;
+
+                case Defines.LEFT:
+                    state.Ego.X = (short)(Defines.MAXX + 1 - state.Ego.XSize);
+                    break;
+            }
+
+            // Change the room number. Note that some games, e.g. MH2, change the CURROOM VAR directly, 
```

Comment:
this comment is now obsolete and can now be deleted

Reply:
Updates and pushed to branch.

Reply:
Done.